Fixing issues with photo verification flow

tasawernawaz opened the pull request:
[ECOM-2979](https://openedx.atlassian.net/browse/ECOM-2979)

* Updated learner dashboard code to show message for verification expiration.
* Updated photo verification view to accept photos if verification is set to expire soon.
* Test added.

**Sandbox:** https://tasawernawaz.sandbox.edx.org/

@ahsan-ul-haq @schenedx Please review this.

schenedx commented on lms/djangoapps/verify_student/models.py:
```diff
@@ -258,6 +258,28 @@ def verification_valid_or_pending(cls, user, earliest_allowed_date=None, queryse
+        if queryset is None:
+            queryset = cls.objects.filter(user=user)
+
+        queryset = queryset.filter(status='approved').order_by('-updated_at')
```
You can use `.first()` method to get the first one of the result. https://docs.djangoproject.com/en/1.10/ref/models/querysets/#first

tasawernawaz replied:
Done.

schenedx commented:
@tasawernawaz In your sandbox, the studio have trouble loading CSS and JS. I could not really use studio and could not verify your changes here in the sandbox. Can you fix the sandbox so I can verify again?

tasawernawaz commented:
@schenedx I have fixed issue with sandbox. I had also added verified mode for edX Demo course. If you don't want to create a new course you can test with edX Demo course.

tasawernawaz commented on lms/djangoapps/verify_student/views.py:
```diff
@@ -1376,12 +1376,23 @@ def get(self, request):
 
+        expiration_datetime = SoftwareSecurePhotoVerification.get_expiration_datetime(request.user)
+        can_reverify = False
+        if expiration_datetime:
```
@schenedx I can't see any other place with the same `if`. Can you please explain your requirement.

tasawernawaz replied:
Done

tasawernawaz commented:
jenkins run lettuce

tasawernawaz commented:
@schenedx Please review changes, I have also updated code sandbox where you can verify.  https://tasawernawaz.sandbox.edx.org/

tasawernawaz commented:
After adding "Current" with statues in sidebar, it looks like this. Do we need to persist with the change and need to update CSS or we should roll beck this change.

<img width="333" alt="screen shot 2016-10-05 at 6 24 05 pm" src="https://cloud.githubusercontent.com/assets/4245618/19115258/ba22b066-8b2a-11e6-8a72-54e5ed2b13fa.png">

tasawernawaz commented:
jenkins run lettuce

schenedx commented:
@rlucioni, can you help review this change? If you need more context, I can talk to you

schenedx commented on common/djangoapps/student/helpers.py:
```diff
@@ -40,6 +41,8 @@ def check_verify_status_by_course(user, course_enrollments):
         * VERIFY_STATUS_NEED_TO_VERIFY: The student has not yet submitted photos for verification.
         * VERIFY_STATUS_SUBMITTED: The student has submitted photos for verification,
           but has have not yet been approved.
+        * VERIFY_STATUS_RESUBMITTED: The student has re-submitted photos for re-verification,
```
`The student has re-submitted photos for re-verification while they still have an active but expiring ID verification`

schenedx commented on common/djangoapps/student/helpers.py:
```diff
@@ -98,6 +105,10 @@ def check_verify_status_by_course(user, course_enrollments):
             relevant_verification = SoftwareSecurePhotoVerification.verification_for_datetime(deadline, verifications)
 
+            verification_expiring_soon = False
+            if SoftwareSecurePhotoVerification.is_verification_expiring_soon(expiration_datetime):
```
@tasawernawaz why can't this be `verification_expiring_soon = SoftwareSecurePhotoVerification.is_verification_expiring_soon(expiration_datetime)`

tasawernawaz replied:
Yes we can. Fixed.

schenedx commented:
@tasawernawaz Some minor comments, rest LGTM. Good work!

tasawernawaz commented:
@schenedx Thanks. Given suggestions are implemented. Please review.

schenedx commented on lms/templates/dashboard/_dashboard_course_listing.html:
```diff
@@ -319,22 +320,24 @@ <h4 class="message-title">${_('Almost there!')}</h4>
+            <h4 class="message-title">${_('You have submitted your verification information.')}</h4>
+            <p class="message-copy">${_('You will see a message on your dashboard when the verification process is complete (usually within 1-2 days).')}</p>
+          % elif verification_status['status'] == VERIFY_STATUS_RESUBMITTED:
+            <h4 class="message-title">${_('You have submitted your reverification information.')}</h4>
```
@tasawernawaz Is this a mistake? I thought the title @srpearce provided for this message is `Your current verification will expire soon!`

tasawernawaz replied:
Sorry Its my bad, I have mixed this up. I will fix in shortly and let you know.

rlucioni commented on common/djangoapps/student/helpers.py:
```diff
@@ -127,11 +143,17 @@ def check_verify_status_by_course(user, course_enrollments):
+                            status = VERIFY_STATUS_NEED_TO_REVERIFY
+                        else:
+                            if verification_expiring_soon:
+                                status = VERIFY_STATUS_RESUBMITTED
```
This part looks strange to me. `verification_expiring_soon = SoftwareSecurePhotoVerification.is_verification_expiring_soon(expiration_datetime)`. So why bother checking `verification_expiring_soon` after checking `SoftwareSecurePhotoVerification.is_verification_expiring_soon(expiration_datetime)`? In fact, why not just use `verification_expiring_soon` on line 147?

rlucioni commented on common/djangoapps/student/tests/test_verification_status.py:
```diff
@@ -202,16 +205,36 @@ def test_verification_will_expire_by_deadline(self):
+        attempt.approve()
+        attempt.save()
+
+        # Expect that learner can submit photos if verification is set to expired soon.
```
`Verify that learner can submit photos if verification is set to expire soon.`

rlucioni commented on common/djangoapps/student/tests/test_verification_status.py:
```diff
@@ -202,16 +205,36 @@ def test_verification_will_expire_by_deadline(self):
-        # Expect that the "verify now" message is hidden
-        # (since the user isn't allowed to submit another attempt while
-        # a verification is active).
+        # Expect that learner can submit photos if verification is set to expired soon.
```
`Verify that learner can submit photos if verification is set to expire soon.`

rlucioni commented on lms/djangoapps/verify_student/models.py:
```diff
@@ -258,6 +258,28 @@ def verification_valid_or_pending(cls, user, earliest_allowed_date=None, queryse
+                of hitting the database.
+
+        Returns:
+            expiration_datetime: expiration_datetime of of most recent "approved"
```
You have an extra "of" in here.

rlucioni commented on lms/djangoapps/verify_student/models.py:
```diff
@@ -258,6 +258,28 @@ def verification_valid_or_pending(cls, user, earliest_allowed_date=None, queryse
     @classmethod
+    def get_expiration_datetime(cls, user, queryset=None):
+        """
+        Check whether the user has an approved verification and Return the
```
s/Return/return

rlucioni commented on common/djangoapps/student/helpers.py:
```diff
@@ -98,6 +105,9 @@ def check_verify_status_by_course(user, course_enrollments):
 
             relevant_verification = SoftwareSecurePhotoVerification.verification_for_datetime(deadline, verifications)
 
+            verification_expiring_soon = SoftwareSecurePhotoVerification.is_verification_expiring_soon(
```
Why is this value assigned inside the loop? It doesn't seem to have anything to do with the enrollment.

rlucioni commented on lms/djangoapps/verify_student/models.py:
```diff
@@ -944,6 +966,18 @@ def verification_status_for_user(cls, user, course_id, user_enrollment_mode):
+        """
+        if expiration_datetime:
+            if (expiration_datetime - datetime.now(pytz.UTC)).days <= settings.VERIFY_STUDENT.get(
+                    "VERIFICATION_EXPIRATION_DAYS", 28):
```
Specifying a default value for a setting in code like this is an anti-pattern. Please don't do this. It leads to surprising behavior. If you need 28 to be the default, make that the default value of the setting.

tasawernawaz commented:
@rlucioni Please review this. Meanwhile I am trying to configure sandbox with latest changes to verify.

tasawernawaz commented on lms/envs/common.py:
```diff
@@ -2283,6 +2283,7 @@
 ################# Student Verification #################
 VERIFY_STUDENT = {
     "DAYS_GOOD_FOR": 365,  # How many days is a verficiation good for?
+    "VERIFICATION_EXPIRATION_DAYS": 28,  # How many days before we show message to reverify?
```
@rlucioni Do I need to add this setting variable in any other setting variable to work on sandboxes.

rlucioni replied:
Not that I'm aware of.

rlucioni replied:
This variable name and the accompanying comment can be clearer. They're vague right now. Clarity around this part of the system is valuable.

The variable represents the window within which a verification is considered to be "expiring soon." (This would make a better comment.)`EXPIRING_SOON_WINDOW` would be a more meaningful variable name. Since this is part of the verification settings dict, you don't need the word "verification" in there.

rlucioni commented on common/djangoapps/student/helpers.py:
```diff
@@ -80,12 +83,18 @@ def check_verify_status_by_course(user, course_enrollments):
 
     recent_verification_datetime = None
 
+    verification_expiring_soon = SoftwareSecurePhotoVerification.is_verification_expiring_soon(expiration_datetime)
```
Put this on line 89. There's no reason to keep it so far from the other line you're adding, which is related.

tasawernawaz commented:
@rlucioni Code updated according to your suggestions, Please review.

tasawernawaz commented:
jenkins run bokchoy
